gammapy/estimators/energydependence.py

```diff
@@ -0,0 +1,245 @@
+import numpy as np
+from gammapy.datasets import Datasets
+from gammapy.modeling import Fit
+from gammapy.modeling.models import FoVBackgroundModel
+from gammapy.modeling.selection import TestStatisticNested
+from gammapy.stats.utils import ts_to_sigma
+from .core import Estimator
+
+
+def weighted_chi2_parameter_results(table_edep, parameter="sigma"):
+    """Calculate the weighted chi2 value for the parameter of interest"""
+
+    values = table_edep[f"{parameter}"][1:]
+    errors = table_edep[f"{parameter}_err"][1:]
+
+    weights = 1 / errors**2
+    avg = np.average(values, weights=weights)
+
+    chi2_value = np.sum((values - avg) ** 2 / errors**2)
+    df = len(values) - 1
+    sigma_value = ts_to_sigma(chi2_value, df)
+
+    chi2_result = {}
+    chi2_result[f"chi2 {parameter}"] = [chi2_value]
+    chi2_result["df"] = [df]
+    chi2_result["significance"] = [sigma_value]
+
+    return chi2_result
+
+
+class EnergyDependenceEstimator(Estimator):
+    """Test if there is any energy-dependent morphology in a map dataset for a given set of energy bins.
+
+    Parameters
+    ----------
+    energy_edges : `~astropy.units.Quantity`
+        Energy edges for the energy-dependence test.
+    source : str or int
+        For which source in the model to compute the estimator.
+    fit : `Fit`
+        Fit instance specifying the backend and fit options.
+
+    """
+
+    tag = "EnergyDependenceEstimator"
+
+    def __init__(self, energy_edges, source, fit=None):
+
+        self.energy_edges = energy_edges
+        self.source = source
+        self.num_energy_bands = len(self.energy_edges) - 1
+
+        if fit is None:
+            fit = Fit(optimize_opts={"print_level": 1})
+
+        self.fit = fit
+
+    def estimate_source_significance(self, datasets):
+        """Estimate the significance of the source above the background.
+
+        Parameters
+        ----------
+        datasets : `~gammapy.datasets.Datasets`
+            Input dataset to use.
+
+        Returns
+        -------
+        table_bkg_src : `~astropy.table.Table`
+            Table with the results of the null hypothesis with no source, and alternative
+            hypothesis with the source added in. Entries are:
+            * "Emin" : the minimum energy of the energy band
+            * "Emax" : the maximum energy of the energy band
+            * "delta_ts" : difference in ts
+            * "df" : the degrees of freedom between null and alternative hypothesis
+            * "significance" : significance of the result
+        """
+        for dataset in datasets:
+            dataset.mask_fit = dataset.counts.geom.energy_mask(
```

Comment:
here, the initial mask of the dataset is erased. Is it the wished behaviour? or an AND is preferable?

Reply:
This was the wished behaviour, however, I am not sure whether it is preferable